File: ios/JustSpent/JustSpent/Services/AutoRecordingCoordinator.swift
import Foundation
import SwiftUI
import Speech
import AVFoundation

/**
 * Auto-Recording Coordinator
 *
 * Orchestrates automatic voice recording when app opens or returns to foreground.
 * Works in conjunction with AppLifecycleManager to ensure proper conditions.
 *
 * Responsibilities:
 * - Verify permissions before auto-recording
 * - Delay recording to allow UI to settle (500ms)
 * - Prevent duplicate recording sessions
 * - Trigger voice recording through ContentView binding
 *
 * Architecture Pattern: Coordinator Pattern
 * - Coordinates between lifecycle, permissions, and recording systems
 * - Single Responsibility: Only handles auto-recording orchestration
 * - Dependency Injection: Receives dependencies through init
 */
@MainActor
class AutoRecordingCoordinator: ObservableObject {

    // MARK: - Published State

    /// Request to start recording (observed by ContentView)
    @Published var shouldStartRecording: Bool = false

    // MARK: - Dependencies

    private let lifecycleManager: AppLifecycleManager

    // MARK: - Configuration

    /// Delay before starting auto-recording (immediate for better UX)
    private let autoRecordingDelay: TimeInterval = 0.0

    // MARK: - Private State

    private var autoRecordingTask: Task<Void, Never>?
    private var isProcessingAutoRecord = false

    // MARK: - Lifecycle

    init(lifecycleManager: AppLifecycleManager) {
        self.lifecycleManager = lifecycleManager

        #if DEBUG
        print("🎤 AutoRecordingCoordinator initialized")
        #endif
    }

    deinit {
        autoRecordingTask?.cancel()
    }

    // MARK: - Public Interface

    /**
     * Attempt to trigger auto-recording
     * Called when app becomes active or returns to foreground
     *
     * - Parameter isRecordingActive: Current recording state from ContentView
     * - Parameter permissions: Current permission state
     */
    func triggerAutoRecordingIfNeeded(
        isRecordingActive: Bool,
        speechPermissionGranted: Bool,
        microphonePermissionGranted: Bool,
        speechRecognitionAvailable: Bool
    ) {
        #if DEBUG
        print("🎤 AutoRecordingCoordinator.triggerAutoRecordingIfNeeded() called")
        print("   - isProcessingAutoRecord: \(isProcessingAutoRecord)")
        print("   - isRecordingActive: \(isRecordingActive)")
        print("   - speechRecognitionAvailable: \(speechRecognitionAvailable)")
        print("   - speechPermissionGranted: \(speechPermissionGranted)")
        print("   - microphonePermissionGranted: \(microphonePermissionGranted)")
        #endif

        // Guard against concurrent auto-recording attempts
        guard !isProcessingAutoRecord else {
            #if DEBUG
            print("⏸️ Auto-recording already processing, skipping")
            #endif
            return
        }

        // Check lifecycle conditions
        guard lifecycleManager.shouldTriggerAutoRecording() else {
            // shouldTriggerAutoRecording() logs reasons internally
            return
        }

        // Check if already recording (manual or previous auto)
        guard !isRecordingActive else {
            #if DEBUG
            print("⏸️ Auto-recording skipped: already recording")
            #endif
            return
        }

        // Verify speech recognition is available
        guard speechRecognitionAvailable else {
            #if DEBUG
            print("⏸️ Auto-recording skipped: speech recognition unavailable")
            #endif
            return
        }

        // Verify all permissions granted
        guard speechPermissionGranted && microphonePermissionGranted else {
            #if DEBUG
            print("⏸️ Auto-recording skipped: permissions not granted")
            if !speechPermissionGranted { print("   - Speech recognition permission missing") }
            if !microphonePermissionGranted { print("   - Microphone permission missing") }
            #endif
            return
        }

        #if DEBUG
        print("✅ All auto-recording conditions met - starting with delay")
        #endif

        // All conditions met - trigger auto-recording with delay
        startAutoRecordingWithDelay()
    }

    /**
     * Cancel any pending auto-recording
     * Call this if user manually triggers recording or app goes to background
     */
    func cancelPendingAutoRecording() {
        autoRecordingTask?.cancel()
        autoRecordingTask = nil
        isProcessingAutoRecord = false

        #if DEBUG
        print("🚫 Pending auto-recording cancelled")
        #endif
    }

    /**
     * Notify coordinator that auto-recording has completed
     * Updates lifecycle manager state
     */
    func autoRecordingDidComplete() {
        lifecycleManager.stopAutoRecording()
        isProcessingAutoRecord = false

        #if DEBUG
        print("✅ Auto-recording completed")
        #endif
    }

    // MARK: - Private Methods

    private func startAutoRecordingWithDelay() {
        // Cancel any existing task
        autoRecordingTask?.cancel()

        isProcessingAutoRecord = true
        lifecycleManager.startAutoRecording()

        #if DEBUG
        print("⏳ Auto-recording scheduled (delay: \(autoRecordingDelay)s)")
        #endif

        autoRecordingTask = Task { @MainActor in
            do {
                // Wait for UI to settle
                try await Task.sleep(nanoseconds: UInt64(autoRecordingDelay * 1_000_000_000))

                // Check if still valid (not cancelled, app still active)
                guard !Task.isCancelled,
                      lifecycleManager.appState == .active else {
                    #if DEBUG
                    print("⏸️ Auto-recording cancelled before execution")
                    #endif
                    isProcessingAutoRecord = false
                    lifecycleManager.stopAutoRecording()
                    return
                }

                // Trigger recording via published property
                #if DEBUG
                print("🎙️ Triggering auto-recording now")
                #endif

                shouldStartRecording = true

                // Reset flag after brief delay to allow observation
                try await Task.sleep(nanoseconds: 100_000_000) // 100ms
                shouldStartRecording = false

                // Note: autoRecordingDidComplete() will be called when recording finishes

            } catch {
                // Task was cancelled
                #if DEBUG
                print("⏸️ Auto-recording task cancelled")
                #endif
                isProcessingAutoRecord = false
                lifecycleManager.stopAutoRecording()
            }
        }
    }

    /**
     * Manual trigger for testing
     */
    #if DEBUG
    func forceAutoRecording() {
        shouldStartRecording = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 100_000_000)
            shouldStartRecording = false
        }
    }
    #endif
}

// MARK: - Permission State Helper

/**
 * Permission state check result
 * Used to determine if auto-recording should proceed
 */
struct PermissionState {
    let speechRecognitionGranted: Bool
    let microphoneGranted: Bool
    let speechRecognitionAvailable: Bool

    var allGranted: Bool {
        speechRecognitionGranted && microphoneGranted && speechRecognitionAvailable
    }

    static var current: PermissionState {
        let speechAuthStatus = SFSpeechRecognizer.authorizationStatus()
        let micPermission = AVAudioSession.sharedInstance().recordPermission
        let recognizer = SFSpeechRecognizer(locale: Locale(identifier: "en-US"))

        return PermissionState(
            speechRecognitionGranted: speechAuthStatus == .authorized,
            microphoneGranted: micPermission == .granted,
            speechRecognitionAvailable: recognizer?.isAvailable ?? false
        )
    }
}

File: android/app/src/main/java/com/justspent/app/voice/AutoRecordingCoordinator.kt
package com.justspent.app.voice

import com.justspent.app.lifecycle.AppLifecycleManager
import com.justspent.app.lifecycle.AppState
import com.justspent.app.permissions.PermissionManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Auto-Recording Coordinator
 *
 * Orchestrates automatic voice recording when app opens or returns to foreground.
 * Works in conjunction with AppLifecycleManager to ensure proper conditions.
 * Mirrors iOS implementation from AutoRecordingCoordinator.swift.
 *
 * Responsibilities:
 * - Verify permissions before auto-recording
 * - Delay recording to allow UI to settle (500ms)
 * - Prevent duplicate recording sessions
 * - Trigger voice recording through StateFlow
 *
 * Architecture Pattern: Coordinator Pattern
 * - Coordinates between lifecycle, permissions, and recording systems
 * - Single Responsibility: Only handles auto-recording orchestration
 * - Dependency Injection: Receives dependencies through Hilt
 */
@Singleton
class AutoRecordingCoordinator @Inject constructor(
    private val lifecycleManager: AppLifecycleManager,
    private val permissionManager: PermissionManager
) {

    companion object {
        private const val TAG = "AutoRecordingCoordinator"
        private const val AUTO_RECORDING_DELAY_MS = 0L // Immediate start for better UX
    }

    // Coroutine scope for auto-recording tasks
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // Request to start recording (observed by Activity/UI)
    private val _shouldStartRecording = MutableStateFlow(false)
    val shouldStartRecording: StateFlow<Boolean> = _shouldStartRecording.asStateFlow()

    // Private state
    private var autoRecordingJob: Job? = null
    private var isProcessingAutoRecord = false

    init {
        android.util.Log.d(TAG, "🎤 AutoRecordingCoordinator initialized")
    }

    /**
     * Attempt to trigger auto-recording
     * Called when app becomes active or returns to foreground
     *
     * @param isRecordingActive Current recording state from UI
     */
    suspend fun triggerAutoRecordingIfNeeded(isRecordingActive: Boolean) {
        // Guard against concurrent auto-recording attempts
        if (isProcessingAutoRecord) {
            android.util.Log.d(TAG, "⏸️ Auto-recording already processing, skipping")
            return
        }

        // Check lifecycle conditions
        if (!lifecycleManager.shouldTriggerAutoRecording()) {
            // shouldTriggerAutoRecording() logs reasons internally
            return
        }

        // Check if already recording (manual or previous auto)
        if (isRecordingActive) {
            android.util.Log.d(TAG, "⏸️ Auto-recording skipped: already recording")
            return
        }

        // Check recording state from VoiceRecordingManager
        // Note: This is checked in UI layer to avoid circular dependencies

        // Verify all permissions granted
        if (!permissionManager.areAllPermissionsGranted()) {
            android.util.Log.d(TAG, "⏸️ Auto-recording skipped: permissions not granted")
            return
        }

        // All conditions met - trigger auto-recording with delay
        startAutoRecordingWithDelay()
    }

    /**
     * Cancel any pending auto-recording
     * Call this if user manually triggers recording or app goes to background
     */
    fun cancelPendingAutoRecording() {
        autoRecordingJob?.cancel()
        autoRecordingJob = null
        isProcessingAutoRecord = false

        android.util.Log.d(TAG, "🚫 Pending auto-recording cancelled")
    }

    /**
     * Notify coordinator that auto-recording has completed
     * Updates lifecycle manager state
     */
    fun autoRecordingDidComplete() {
        lifecycleManager.stopAutoRecording()
        isProcessingAutoRecord = false

        android.util.Log.d(TAG, "✅ Auto-recording completed")
    }

    /**
     * Start auto-recording with delay
     * Private method that handles the actual triggering
     */
    private fun startAutoRecordingWithDelay() {
        // Cancel any existing job
        autoRecordingJob?.cancel()

        isProcessingAutoRecord = true
        lifecycleManager.startAutoRecording()

        android.util.Log.d(TAG, "⏳ Auto-recording scheduled (delay: ${AUTO_RECORDING_DELAY_MS}ms)")

        autoRecordingJob = scope.launch {
            try {
                // Wait for UI to settle
                delay(AUTO_RECORDING_DELAY_MS)

                // Check if still valid (not cancelled, app still active)
                val appState = lifecycleManager.appState.first()
                if (appState != AppState.ACTIVE) {
                    android.util.Log.d(TAG, "⏸️ Auto-recording cancelled: app not active")
                    isProcessingAutoRecord = false
                    lifecycleManager.stopAutoRecording()
                    return@launch
                }

                // Trigger recording via StateFlow
                android.util.Log.d(TAG, "🎙️ Triggering auto-recording now")

                _shouldStartRecording.value = true

                // Reset flag after brief delay to allow observation
                delay(100L) // 100ms
                _shouldStartRecording.value = false

                // Note: autoRecordingDidComplete() will be called when recording finishes

            } catch (e: Exception) {
                // Job was cancelled or error occurred
                android.util.Log.d(TAG, "⏸️ Auto-recording task cancelled or failed: ${e.message}")
                isProcessingAutoRecord = false
                lifecycleManager.stopAutoRecording()
            }
        }
    }

    /**
     * Manual trigger for testing
     */
    fun forceAutoRecordingForTesting() {
        scope.launch {
            _shouldStartRecording.value = true
            delay(100L)
            _shouldStartRecording.value = false
        }
    }

    /**
     * Clean up resources
     */
    fun cleanup() {
        autoRecordingJob?.cancel()
        autoRecordingJob = null
        isProcessingAutoRecord = false
        android.util.Log.d(TAG, "🧹 AutoRecordingCoordinator cleaned up")
    }
}
